include/dmlc/thread_group.h

```diff
@@ -395,38 +404,46 @@ class ThreadGroup {
    */
   inline void join_all() {
     CHECK_EQ(!is_this_thread_in(), true);
-    std::unordered_set<std::shared_ptr<Thread>> working_set;
-    {
-      ReadLock guard(m_);
-      for (auto iter = threads_.begin(), e_iter = threads_.end(); iter != e_iter; ++iter) {
-        if (!(*iter)->is_auto_remove()) {
-          working_set.emplace(*iter);
+    do {
```

Comment:
Here is just a scoped addition of td::unique_lock<std::mutex> lk(join_all_mtx_);